config/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurer.java

```diff
@@ -0,0 +1,228 @@
+ *
+ * <ul>
+ * <li>{@link AuthenticationManager}</li>
+ * </ul>
+ *
+ * If {@link #jwt()} isn't supplied, then the {@link BearerTokenAuthenticationFilter} is still added, but without
+ * any OAuth 2.0 {@link AuthenticationProvider}s. This is useful if needing to switch out Spring Security's Jwt support
+ * for a custom one.
+ *
+ * @author Josh Cummings
+ * @since 5.1
+ * @see BearerTokenAuthenticationFilter
+ * @see JwtAuthenticationProvider
+ * @see NimbusJwtDecoderJwkSupport
+ * @see AbstractHttpConfigurer
+ */
+public final class OAuth2ResourceServerConfigurer<H extends HttpSecurityBuilder<H>> extends
+		AbstractHttpConfigurer<OAuth2ResourceServerConfigurer<H>, H> {
+
+	private BearerTokenResolver resolver = new DefaultBearerTokenResolver();
+	private BearerTokenRequestMatcher matcher = new BearerTokenRequestMatcher();
+
+	private BearerTokenAuthenticationEntryPoint entryPoint = new BearerTokenAuthenticationEntryPoint();
+	private BearerTokenAccessDeniedHandler deniedHandler = new BearerTokenAccessDeniedHandler();
+
+	private JwtConfigurer jwtConfigurer;
+
+	public JwtConfigurer jwt() {
+		if (this.jwtConfigurer == null) {
+			this.jwtConfigurer = new JwtConfigurer();
+		}
+
+		return this.jwtConfigurer;
+	}
+
+	@Override
+	public void setBuilder(H http) {
+		super.setBuilder(http);
+		initSessionCreationPolicy(http);
+	}
+
+	@Override
+	public void init(H http) throws Exception {
+		registerDefaultAccessDeniedHandler(http);
+		registerDefaultEntryPoint(http);
+		registerDefaultCsrfOverride(http);
+	}
+
+	@Override
+	public void configure(H http) throws Exception {
+		BearerTokenResolver resolver = getBearerTokenResolver(http);
+		this.matcher.setResolver(resolver);
+
+		AuthenticationManager manager = http.getSharedObject(AuthenticationManager.class);
+
+		BearerTokenAuthenticationFilter filter =
+				new BearerTokenAuthenticationFilter(manager);
+		filter.setBearerTokenResolver(resolver);
+		filter = postProcess(filter);
+
+		http.addFilterBefore(filter, BasicAuthenticationFilter.class);
+
+		if (this.jwtConfigurer != null) {
+			JwtDecoder decoder = this.jwtConfigurer.getJwtDecoder(http);
+
+			if (decoder != null) {
+				JwtAuthenticationProvider provider =
+						new JwtAuthenticationProvider(decoder);
+				provider = postProcess(provider);
+
+				http.authenticationProvider(provider);
+			} else {
+				throw new BeanCreationException("Jwt is the only supported format for bearer tokens " +
+						"in Spring Security and no instance of JwtDecoder could be found. Make sure to specify " +
+						"a jwk set uri by doing http.oauth2().resourceServer().jwt().jwkSetUri(uri)");
+			}
+		}
+	}
+
+	public class JwtConfigurer {
+		private JwtDecoder decoder;
+
+		JwtConfigurer() {}
+
+		public OAuth2ResourceServerConfigurer<H> jwkSetUri(String uri) {
+			this.decoder = new NimbusJwtDecoderJwkSupport(uri);
+			return OAuth2ResourceServerConfigurer.this;
+		}
+
+		protected JwtDecoder getJwtDecoder(H http) {
```

Comment:
Impressive catch! I actually intentionally left this because I see this in other places in the code, which I figured was meant to signal to others that this variable is an allowable mechanism for making the implementation decision.

Such is the case, for example, in ExceptionHandlingConfigurer already: https://github.com/spring-projects/spring-security/blob/master/config/src/main/java/org/springframework/security/config/annotation/web/configurers/ExceptionHandlingConfigurer.java#L194

Am I reading too much into that (note that I followed it too in ExceptionHandlingConfigurer when I created createDefaultDeniedHandler)?